scheduler/src/cook/mesos/fenzo.clj

```diff
@@ -0,0 +1,84 @@
+;;
+;; Copyright (c) Two Sigma Open Source, LLC
+;;
+;; Licensed under the Apache License, Version 2.0 (the "License");
+;; you may not use this file except in compliance with the License.
+;; You may obtain a copy of the License at
+;;
+;;  http://www.apache.org/licenses/LICENSE-2.0
+;;
+;; Unless required by applicable law or agreed to in writing, software
+;; distributed under the License is distributed on an "AS IS" BASIS,
+;; WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+;; See the License for the specific language governing permissions and
+;; limitations under the License.
+;;
+(ns cook.mesos.fenzo
+  (:require [clojure.tools.logging :as log]
+            [datomic.api :as d])
+  (import java.util.concurrent.TimeUnit
+          org.apache.mesos.Protos$Offer
+          com.netflix.fenzo.TaskAssignmentResult
+          com.netflix.fenzo.TaskScheduler
+          com.netflix.fenzo.VirtualMachineLease
+          com.netflix.fenzo.plugins.BinPackingFitnessCalculators))
+
+(defn extract-message
+  "For some reason, Fenzo's AssignmentFailure doesn't have a getter for the
+  message field.  Need to parse it from the string representation of the
+  AssignmentFailure."
+  [resource-failure]
+  (last (re-find (re-pattern "message=(\\w+)") (str resource-failure))))
+
+(defn count-resource-failure
+  [totals resource-failure]
+  "Fenzo TaskAssignmentResults tend to provide AssignmentFailures in pairs for each
+   resource placement issue. One AssignmentFailure will describe the resource in the
+   message field; the other will use the Resource field.  We actually want to use
+   the message field because the Enumeration of the Resource field doesn't allow
+   for GPU, whereas the message field can contain the string \"gpu\"."
+  (let [msg (extract-message resource-failure)]
+    (if (seq msg)
+      (update-in totals [:resources msg] (fnil inc 0))
+      totals
+      )))
+
+(defn summarize-placement-failure
+  [totals assignment-result]
+  (let [resources-counted (reduce count-resource-failure totals (.getFailures assignment-result))
+        constraint-failure (.getConstraintFailure assignment-result)]
+    (if constraint-failure
+      (update-in resources-counted [:constraints (.getName constraint-failure)] (fnil inc 0))
+      resources-counted)))
+
+(defn accumulate-placement-failures
+  "Param should be a collection of Fenzo placement failure results for a single
+  job.  Returns a tuple of [job id, single string containing all of the result
+  messages].  Along the way, logs each result message if debug-level logging
+  is enabled."
+  [failures]
+  (let [job (:job  (.. (first failures) (getRequest)))
+        under-investigation? (:job/under-investigation job)
+        summary (when (or under-investigation? (log/enabled? :debug))
+                  (reduce summarize-placement-failure {} failures))]
+    (when (log/enabled? :debug)
```

Comment:
Its purpose is to avoid doing the work of constructing the logged string unless the debug threshold is actually enabled.  Probably premature optimization;  the potentially heavy duty calculation is already avoided by another check for (log/enabled? :debug) above...  👍